### PR TITLE
Provider call context and its implementation placeholders.

### PR DESCRIPTION
**Pull request by anastasiamac**
## Description of change

We have found ourselves in a need to react to certain events and occurrences during the calls to cloud providers. For example, we want to keep juju status uptodate or we want to invalidate a cloud credential that we have used to make a call... The cleanest way to achieve this is to make a cloud call with these behaviors defined prior to making the actual call. For example, since we know what we want to happen if the cloud rejects a credential, we want to define it and inject that behavior into the call.

So it has been decided that we want to define a 'call context' that will contain all these useful behaviors as callbacks. The first real callback to be defined in the context will deal with cloud credentials, following closely by status update callback...

This PR is the first in a series and is largely mechanical - it introduces the concept of a call context (see environs/context package), puts it in the signature call on most of the cloud calls as defined by interfaces in environs package and creates a placeholder implementation in most packages that make these cloud calls - cmd/juju and cmd/jujud; apiserver; workers and state.

Note that some call contexts may be created in a slightly different locations than what is proposed here. These details will be ironed out as the implementation is refined.

## QA steps

This PR is not changing any functionality nor facade signatures. Thus, all unit/CI/manual tests should pass as previously. I have successfully bootstrapped, deployed and destroyed AWS models.

### CODE REVIEW

File: apiserver/common/environ_config.go [hunk 1]
```diff
@@ -24,3 +27,19 @@ func (f EnvironConfigGetterFuncs) ModelConfig() (*config.Config, error) {
...
+func ProviderCallContext() context.ProviderCallContext {
+	return &ProviderContext{}
+}
+
+// ProviderContext contains cloud provider call context for provider calls from
+// within apiserver layer.
+type ProviderContext struct{}
+
+// InvalidateCredentialCallback implements ProviderCallContext.InvalidateCredentialCallback.
+func (*ProviderContext) InvalidateCredentialCallback() error {
```
[the remark this answers is not in the document]
Comment (anastasiamac) [reply apiserver/common/environ_config.go]: So this is a placeholder to get the PR to land. The actual implementation of the call will call an underlying func which will have different content based on where the call is made from. For example, if you making a provider call from apiserver used by an agent, the impl might contain a call to a different facade that invalidates credential; where as if you are calling from a state layer, the call will have a state function that will invalidate the credential. The actual implementation depends on where you are when you are making the call and what information you have - do you have access to a model and, hence, need to navigate through it to get to a credential on the controller, do you have it in some parameters, etc...

The context is specific to a call not an object on which the call is made. Hence, the implementation will differ based on where the call is made from...

File: apiserver/common/environ_config.go [hunk 2]
```diff
@@ -24,3 +27,19 @@ func (f EnvironConfigGetterFuncs) ModelConfig() (*config.Config, error) {
...
+// ProviderCallContext returns the context with all necessary functionality,
+// including call backs, to make a call to a cloud provider.
+// TODO (anastasiamac 2018-04-30) make it real
+func ProviderCallContext() context.ProviderCallContext {
+	return &ProviderContext{}
+}
+
+// ProviderContext contains cloud provider call context for provider calls from
+// within apiserver layer.
+type ProviderContext struct{}
```
[the remark this answers is not in the document]
Comment (anastasiamac) [reply apiserver/common/environ_config.go]: Agreed. I'll change to have one context that will essentially, eventually have internal variables as functions.

**Comment by babbageclunk** [comment]
It turns out I was wrong about shared environs - looking at the environ-tracker worker for example, it hands out the same environ to any dependent. So we can't hang call-specific state off it.

**Comment by anastasiamac** [comment]
!!build!!

**Comment by anastasiamac** [comment]
$$merge$$

**Comment by anastasiamac** [comment]
unrelated failure in github.com/juju/juju/worker/raft/raftflag

**Comment by anastasiamac** [comment]
$$merge$$

**Pull request merged**